File: 2023/day4/golang/solution.go
package main

import (
	"fmt"
	"math"
	"strings"
	"time"

	"golang.org/x/exp/slices"
)

func partOne(lines []string) int {
	result := 0

	for _, line := range lines {
		local := -1

		numbers := strings.Split(strings.Split(line, ":")[1], "|")

		winning := slices.DeleteFunc(strings.Split(numbers[0], " "), func(e string) bool {
			return e == ""
		})
		having := slices.DeleteFunc(strings.Split(numbers[1], " "), func(e string) bool {
			return e == ""
		})

		for _, h := range having {
			if slices.Contains(winning, h) {
				local += 1
			}
		}

		if local > -1 {
			result += int(math.Pow(2, float64(local)))
		}
	}

	return result
}

func partTwo(lines []string) int {
	startTime := time.Now()
	d := make(map[int]int, len(lines))

	for i := range lines {
		d[i] = 0
	}

	for lineIdx, line := range lines {
		for i := 0; i <= d[lineIdx]; i++ {
			local := 0

			numbers := strings.Split(strings.Split(line, ":")[1], "|")

			winning := slices.DeleteFunc(strings.Split(numbers[0], " "), func(e string) bool {
				return e == ""
			})
			having := slices.DeleteFunc(strings.Split(numbers[1], " "), func(e string) bool {
				return e == ""
			})

			for _, h := range having {
				if slices.Contains(winning, h) {
					local += 1
				}
			}

			if local > 0 {
				for j := lineIdx + 1; j < lineIdx+1+local; j++ {
					d[j] += 1
				}
			}
		}
	}

	sum := 0

	for _, v := range d {
		sum += v
	}

	duration := time.Since(startTime)
	fmt.Println("time taken: ", duration)
	return sum + len(lines)
}

File: 2023/day4/python/solution.py
from typing import List
import os

def part1(lines: List[str]):
    result = 0
    for line in lines:
        local_result = -1

        numbers = line.strip().split(":")[1].strip().split('|')
        winning = list(filter(len, numbers[0].strip().split(" ")))
        having = list(filter(len, numbers[1].strip().split(" ")))

        for h in having:
            if h in winning:
                local_result += 1
        
        if local_result > -1:
            result += 2 ** (local_result)

    return result

def part2(lines: List[str]):
    d = {}

    for i in range(0, len(lines)):
        d[i] = {"wins": 0, "count": 1}

    for line_idx, line in enumerate(lines):        
        local_result = 0
        numbers = line.strip().split(":")[1].strip().split('|')
        winning = list(filter(len, numbers[0].strip().split(" ")))
        having = list(filter(len, numbers[1].strip().split(" ")))

        for h in having:
            if h in winning:
                local_result += 1

        d[line_idx]["wins"] = local_result
    
    s = 0

    for key, value in d.items():
        for x in range(key+1, key + 1 + value["wins"]):
            d[x]["count"] += 1 * value["count"]
        s += value["count"]

    return s

if __name__ == "__main__":
    f = open(os.getcwd() + '/day4/inputs/main.txt', 'r')
    lines = f.readlines()
    print(part2(lines))